### ISSUE TITLE
Search: index from children nodes up to parent nodes

### ISSUE THREAD

**Issue by stsewd** [readthedocs/readthedocs.org#10184]
When indexing content, we don't want content from nested sections/description lists to overlap/duplicate, so we inspect their content recursively up to 2 levels.

> Sections can be contained in up to two nested tags

https://dev.readthedocs.io/en/latest/search-integration.html#sections

We can avoid having this restriction by changing how we index nodes. Instead of doing a sequential parsing, we start all leaf nodes, and removing them once we parse them, then repeat this again for the new leaf nodes, and so on till we get to the parent nodes.

I don't think there is an CSS selector for this, so we would need to do this manually, a simple/inefficient solution would be to get all leaf nodes on every iteration, a more efficient solution could be constructing a tree with all the section/list nodes, and then depth-first traverse/parse that tree (an probably maybe there is no need to create another tree, and we can just re-use the current one).

**Comment by benjaoming** [comment]
I agree with the intention here, if it can work: It makes the indexes smaller and saves the user time by only giving them the most specific link.

But I do have one search-technical question about the approach.

## The Example

I get that we can always have a way to find the right page for the user if we combine ALL the content in the search index for a page.

I'm wondering if we can find an accurate section, though, or if we might mess it up?

Let's say that we have the following content:

```markdown
# Programming languages

## Ruby tutorial

Welcome to the tutorial

...

## Python tutorial

Welcome to the tutorial

<dl>
<dt id="strings">Strings</dt>
<dd>Strings are types and you can .join() them</dd>
<dt id="integers">Integers</dt>
<dd>Integers are great for numbers</dd>
</dl>
```

## Index: Everything for every section

```
{
  "title": "Programming languages",
  "path": "university-of-everything/programming-languages",
  "sections": [
    {
      "title": "Ruby tutorial",
      "content: "Welcome to the tutorial",
      "id": "ruby-tutorial",
    },
    {
      "title": "Python tutorial",
      "content: "Welcome to the tutorial\nStrings are types and you can .join() them\nIntegers are great for numbers",
      "id": "python-tutorial",
    },
    {
      "title": "Strings",
      "content: "Strings are types and you can .join() them",
      "id": "strings",
    },
    {
      "title": "Integers",
      "content: "Integers are great for numbers",
      "id": "integers",
    },
```


## Index: Everything divided

```
{
  "title": "Programming languages",
  "path": "university-of-everything/programming-languages",
  "sections": [
    {
      "title": "Ruby tutorial",
      "content: "Welcome to the tutorial",
      "id": "ruby-tutorial",
    },
    {
      "title": "Python tutorial",
      "content: "Welcome to the tutorial",
      "id": "python-tutorial",
    },
    {
      "title": "Strings",
      "content: "Strings are types and you can .join() them",
      "id": "strings",
    },
    {
      "title": "Integers",
      "content: "Integers are great for numbers",
      "id": "integers",
    },
```

## The questions

Now when we have the flat search index with divided contents, then we cannot tell anymore that the "strings" section belongs to "Python". So in my mind, there is no way for the search process nor the layout of search results to figure this out?

It's like... What happens if I search using terms that are relevant across multiple sections:

* `python strings`
* `python integers`
* `ruby numbers` (let's say that the ruby section doesn't have the word 'numbers' in it)

## Solution, not sure?

My approach would be to ask if we want to allow duplicate content for all the sections but then have a process that decides which search result is the most _specific_ one. A simple measurement would be if we could choose the one with the _shortest_ contents, since that's the most exact hit.

**Comment by benjaoming** [comment]
A good example that @stsewd uses is this one: https://requests.readthedocs.io/en/latest/api/#requests.Response

**Comment by benjaoming** [comment]
Link to @stsewd's comment about search analyzer: https://github.com/readthedocs/readthedocs.org/pull/10128#pullrequestreview-1377992504 - if we end up doing experiments on different kinds of section layouts, we might also include the search analyzer since it might affect it (or might not).
